[ENH] Parallel processing support for epoched data

ryanhammonds opened the pull request:
This PR is related to #32 and #52. Epoched data, or any group of signals where the same compute_features kwargs should be applied, can now be processed in parallel using the same set of parameters. The implementation here is similar to fitting a FOOOFGroup object.

Update: This PR, so far, is for running 2d arrays of signals (n_epochs, n_points) in parallel. I went back and re-read Natalie's comment and I need to also add support for 3D arrays (n_channels, n_epochs, n_points) where a different set of kwargs is allowed for each channel.

TomDonoghue commented on bycycle/group/epochs.py:
```diff
@@ -0,0 +1,93 @@
+        Keyword arguments used in :func:`~.compute_features`.
+    return_samples : bool, optional, default: True
+        Whether to return a dataframe of cyclepoint sample indices.
+    n_cpus : int, optional, default: -1
```
I think I vote for this param to be called `n_jobs`, mostly just because that's the convention I'm familiar with across other modules, and would be consistent with, for example, FOOOF. 

I _think_ also that perhaps the number of processes is also perhaps not strictly necessarily the same as n_cpus (?).

TomDonoghue commented on bycycle/group/epochs.py:
```diff
@@ -0,0 +1,93 @@
+        Whether to return a dataframe of cyclepoint sample indices.
+    n_cpus : int, optional, default: -1
+        The number of jobs, one per cpu, to compute features in parallel.
+    progress: str, optional, default: None
```
This doesn't take any string, but a set of distinct options, the convention for which is to list them out. 
I think it should be listed as:
{None, 'tqdm', 'tqdm.notebook'} (or similar).

TomDonoghue commented on bycycle/group/utils.py:
```diff
@@ -0,0 +1,64 @@
+    ----------
+    iterable : list or iterable
+        Iterable object to potentially apply progress tracking to.
+    progress : {None, 'tqdm'}, optional
```
I think this is missing the option for 'tqdm.notebook'.

TomDonoghue commented on bycycle/tests/group/test_epochs.py:
```diff
@@ -0,0 +1,60 @@
+        assert df_samples.equals(samples[0])
+
+    # Assert that two cpus computes faster than one
+    assert elapsed_time_parallel < elapsed_time_sequential
```
I'm not so sure about doing the whole timing thing, and testing it. 
Couple reasons:
- this feels to me like profiling, and not really testing
- parallel code is not guaranteed to be faster (sometimes the overhead exceeds gains), and this feels like something that could occur with our small test signals - meaning I wouldn't be surprised if in some possible cases the time check failed, but that would not be an informative / useful failure. 

I think then I would vote to not do the timing checks in these tests. What do you think?

ryanhammonds replied:
I'll remove the time tests. I agree with the second point, I had to increase the size of the array so it would pass.

TomDonoghue commented on bycycle/tests/group/test_utils.py:
```diff
@@ -0,0 +1,21 @@
+
+    pbar = progress_bar(iterable, progress, n_iterations)
+
+    assert type(pbar) == list
```
This test fails for me locally. I currently have tqdm version 4.47.0. 

I get:
`AssertionError: assert <class 'tqdm.std.tqdm'> == list where <class 'tqdm.std.tqdm'>`

I _think_ this is actually right - when `tqdm` is used, then `pbar` is not type list, and so we need to update this test. 

I think it doesn't fail on travis but we haven't added `tqdm` to any dependencies, so it doesn't actually get run using a progress bar?

Maybe though there is some kind of version difference somewhere. Does this test pass for you locally?

ryanhammonds replied:
This needs to be updated. It passes for me locally only when I don't have tqdm installed in my environment. I'll fix this.

TomDonoghue commented:
Thanks for the work here @ryanhammonds!

I did a direct edit of some small lints (things that are annoying to comment), and did a proper review. 

One slightly broader Q: I wonder about using the term `epochs`. The implementation is general to any 2D array (applying either the same, or different settings). It seems it could be applied to cases that aren't necessarily 'epochs'. What do we think about just referring to what we are running here as a 'group' (analogous to how we use the term n FOOOF?).

ryanhammonds commented:
Hi @TomDonoghue, thanks for the review! I push a commit addressing your comments (fb459c7). I renamed `bycycle.group.epochs` to a more general `bycycle.group.features` and fixed a bug in 2b1adf1. And added 3d support in 564b8e7.

Circleci is failing for these updates although it's passing locally. I'll dig into this tomorrow.

TomDonoghue commented:
Just a quick thought on the 3d function: what do we think about flattening and running a single 2d array, versus running a series of 2d arrays (basically `for 2d_slice in 3d_array: process(2d_slice)` - analogous to how FOOOF fits 3d arrays (https://fooof-tools.github.io/fooof/_modules/fooof/objs/utils.html#fit_fooof_3d)

It feels to me like the running across 2d arrays is potentially simpler, and might be easier to manage settings, if they vary across 2d slices?

Also, for DFs that contain results from multiple signals - how do the DF's label the different signals? This feels like it could start to get a little messy with 3d arrays.

ryanhammonds commented:
I think flattening first may be more efficient (in terms of computation time), but I'll double check in a notebook. I think iterating over a 3d array, like in fit_fooof_3d, will result in the opening/closing multiprocessing pools repeatedly for each iteration. Rather than opening/closing single pool. I'm not sure if this matter or how much more efficient it is but I'll check.

Settings are currently managed by passing in a: 
- single dictionary (i.e. same params across all signal).
- 1d list of dicts to apply across the first dimension (i.e. a list of 3 dicts to apply across 2D slices for data of shape (3, X, X)
- 2d list of dicts to apply unique settings for each each signal or 1d array (i.e list of dict shape of (3, 2) for data of shape (3, 2, X).

I'm not sure how iterating over 2d arrays could simplify this, but I'll dig more into how it's done in fooof.

For the dataframe organization, the dictionaries are returned in the same shape as the first two dimensions of the signal. So if a (3, 2, N) array is used, a 2d list of dataframe is returned with a shape (3, 2). That way dataframes may be easily index as `df_features[0][0]` for the first array or `df_features[2][1]` for the last array, again assuming the (3, 2, N) array.

ryanhammonds commented:
I tested iterating over 2d arrays versus flattening into a single 2d array, and flattening runs faster. With a 3d array of shape (20, 20, 5000), flattening took 1.83s versus 2.99s when iterating (both using n_jobs=12). The time difference of ~1s doesn't seem like a lot, but it scales with more participants and longer signals.

ryanhammonds commented:
I start working on updating documentation for 1.0 and though I'd leave example usage of compute_features_3d here to show the different ways it can work.
```
from bycycle.group import compute_features_3d
import numpy as np
from neurodsp.sim import sim_bursty_oscillation

# Simulate a 3d array of shape (n_channels, n_epochs, n_signals)
fs, f_range, n_chs, n_epochs = 500, (8, 12), 2, 2
sigs = np.array([[sim_bursty_oscillation(10, fs, 10) for epoch in range(n_epochs)] for ch in range(n_chs)])


# Apply the same thresholds across all channels and epochs
threshold_kwargs = {'amp_consistency_threshold': .5, 'period_consistency_threshold': .5,
                    'monotonicity_threshold': .8, 'min_n_cycles': 3}

features = compute_features_3d(sigs, fs, f_range, return_samples=False, n_jobs=2,
                               compute_features_kwargs={'threshold_kwargs': threshold_kwargs})


# Apply channel-specific thresholds
threshold_kwargs_ch1 = {'amp_consistency_threshold': .25, 'period_consistency_threshold': .25,
                        'monotonicity_threshold': .25, 'min_n_cycles': 3}

threshold_kwargs_ch2 = {'amp_consistency_threshold': .5, 'period_consistency_threshold': .5,
                        'monotonicity_threshold': .5, 'min_n_cycles': 3}

compute_kwargs = [{'threshold_kwargs': threshold_kwargs_ch1}, {'threshold_kwargs': threshold_kwargs_ch1}]

features = compute_features_3d(sigs, fs, f_range, return_samples=False, n_jobs=2,
                               compute_features_kwargs={'threshold_kwargs': compute_kwargs})


# Apply unique threshold for each channel and epoch
threshold_kwargs_ch1_epoch1 = {'threshold_kwargs': 
                                   {'amp_consistency_threshold': .2, 'period_consistency_threshold': .2,
                                    'monotonicity_threshold': .2}
                              }

threshold_kwargs_ch1_epoch2 = {'threshold_kwargs':  
                                   {'amp_consistency_threshold': .4, 'period_consistency_threshold': .4,
                                    'monotonicity_threshold': .4}
                              }

threshold_kwargs_ch2_epoch1 = {'threshold_kwargs': 
                                   {'amp_consistency_threshold': .6, 'period_consistency_threshold': .6,
                                   'monotonicity_threshold': .6}
                              }

threshold_kwargs_ch2_epoch2 = {'threshold_kwargs': 
                                   {'amp_consistency_threshold': .8, 'period_consistency_threshold': .8,
                                    'monotonicity_threshold': .8}
                              }

compute_kwargs = [[threshold_kwargs_ch1_epoch1, threshold_kwargs_ch1_epoch2],
                  [threshold_kwargs_ch2_epoch1, threshold_kwargs_ch2_epoch2]]

features = compute_features_3d(sigs, fs, f_range, return_samples=False, n_jobs=2,
                               compute_features_kwargs=compute_kwargs)
```

TomDonoghue commented:
I left a bunch of relatively small comments in the proper review above. 

In terms of the approach for looking into it and checking some profiling. What you said & propose makes sense to me - lets keep it as is. 

In terms of the code and overall approach (other than the small comments), this all looks good to me. I think this is the kind of thing that we can merge into master soon, and hopefully get some test users on it, just to make sure the organization and usage is smooth. I wonder if, for example, it might be useful to offer a helper function that could collapse across collections of dataframes, adding row that indicates which signal each row comes from - but deciding on that kind of thing will probably become clearer with usage tests.

ryanhammonds commented on bycycle/tests/group/test_utils.py:
```diff
@@ -0,0 +1,21 @@
+"""Test group utility functions."""
+
+import time
```
Whoops, yeah. Updating this.

ryanhammonds commented on bycycle/group/features.py:
```diff
@@ -0,0 +1,228 @@
+    elif compute_features_kwargs is None:
+        compute_features_kwargs = {}
+
+    # Remove return_samples from compute_features_kwargs
```
Thanks for this! I totally agree. I went with this to keep line lengths shorter and to avoid aliases:
```
compute_features_kwargs = [compute_features_kwargs] if \
    isinstance(compute_features_kwargs, dict) else compute_features_kwargs
[kwargs.pop('return_samples', None) for kwargs in compute_features_kwargs]
```

ryanhammonds commented on bycycle/group/utils.py:
```diff
@@ -0,0 +1,64 @@
+    if progress:
+
+        # Try loading the tqdm module
+        try:
```
Thanks for the tip! That worked:
![tqdm](https://user-images.githubusercontent.com/34786005/94313299-4ee0ed00-ff33-11ea-90d0-9755874801ea.png)

ryanhammonds commented on bycycle/tests/group/test_features.py:
```diff
@@ -0,0 +1,136 @@
+"""Test functions to compute features across epoched data."""
+
+import time
```
Yep, dropping!

ryanhammonds commented on bycycle/tests/group/test_features.py:
```diff
@@ -0,0 +1,136 @@
+    fs  = sim_args['fs']
+    f_range = sim_args['f_range']
+
+    # return_samples is disregarded when used in compute_features_kwargs.
```
Updating.

Which comments? I don't see one here about this, but I remember you left a comment periods in comments in one of Eric's PRs a week or so ago. I typically don't use periods either, not that it really matters though.

TomDonoghue replied:
hahaha, the second line was a joke, as in "did you miss me be overly specific with a ton of comments". :P

I did not at all mean to imply you had not noticed some genuine comment somewhere :)

ryanhammonds replied:
Ohhhh, yeah totally misread that lol.

I'm going to compile a list of your nitpicks, and write a script to parse code and fix those nitpicks automatically. It can one of those automated PR check things :wink:

ryanhammonds commented on bycycle/tests/group/test_features.py:
```diff
@@ -0,0 +1,136 @@
+    features, samples = compute_features_2d(sigs, fs, f_range, n_jobs=-1, return_samples=True,
+                                            compute_features_kwargs=compute_features_kwargs)
+
+    # Since the same signal is used, check that each df is the same
```
Updated for this. Thanks!

ryanhammonds commented:
I pushed an update addressing you comments. I think this is ready to merge. Unless you think the helper function belongs here.

A helper function to collapse dataframes would be useful I think. I can include that in a a future PR after some test users have played with these funcs, or here.
